R/SandwichLayerVariance.R

```diff
@@ -558,3 +559,459 @@ vcov_tee <- function(x, type = "CR0", cluster = NULL, ...) {
+  name_blk <- colnames(design_obj@structure)[design_obj@column_index == "b"]
+  name_clu <- colnames(design_obj@structure)[design_obj@column_index == "u"]
+  
+  data_temp <- .merge_block_id_cols(data_temp, name_blk)
+  name_blk <- name_blk[1]
+  
+  eq_clu <- paste(name_clu, collapse = " + ")
+  form <- paste("cbind(", name_trt, ", ", name_blk, ") ~ ", eq_clu, sep = "")
+  form2 <- paste("cbind(.wy, .w, .w0) ~ ", eq_clu, sep = "")
+  
+  # aggregate z and bid by mean, aggregate w, w0, and w*y by sum
+  data_aggr <- cbind(
+    do.call("aggregate", list(as.formula(form), FUN = mean, data = data_temp)),
+    do.call("aggregate", list(as.formula(form2), FUN = sum, data = data_temp))
+  )
+  data_aggr$.wy <- data_aggr$.wy / data_aggr$.w
+  data_aggr <- data_aggr[order(data_aggr[, name_blk]), ]
+  return(list(data = data_aggr, block = name_blk, z = name_trt))
+}
+
+#' @title (Internal) Merge multiple block IDs
+#' @param df a data frame 
+#' @param ids a vector of block IDs, column names of df
+#' @details merge multiple block ID columns by the value combinations
+#' and store the new block ID in the column \code{ids[1]}
+#' @return a data frame with a column that contains unique block number IDs
+#' @keywords internal
+.merge_block_id_cols <- function(df, ids){
+  df[,ids[1]] <- apply(df[, ids, drop = FALSE], 1, paste, collapse = "_")
+  unique_ids <- data.frame(unique(df[,ids[1]]))
+  colnames(unique_ids) <- ids[1]
+  unique_ids$.ID <- seq_len(nrow(unique_ids))
+  
+  df <- merge(df, unique_ids, by = ids[1], all.x = TRUE)
+  df[,ids[1]] <- df$.ID
+  df$.ID <- NULL
+  return(df)
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @param x a fitted \code{teeMod} model
+#' @keywords internal
+.prepare_design_matrix <- function(x, ...) {
+  design_obj <- x@Design
+  data <- x$call$data
+  name_clu <- colnames(design_obj@structure)[design_obj@column_index == "u"]
+  cid <- .merge_block_id_cols(data, name_clu)[, name_clu[1]]
+  
+  agg <- .aggregate_to_cluster(x)
+  bid <- agg[[1]][, agg[[2]]]  # block ids
+  
+  name_y <- as.character(x$terms[[2]]) # name of the outcome column
+  X1 <- model.matrix(x$call$offset@fitted_covariance_model) # design matrix
+  p <- ncol(X1)
+  
+  wc <- x$call$offset@fitted_covariance_model$weight
+  if (is.null(wc)) wc <- 1
+  X1 <- matrix(
+    rep(wc * (data[, name_y] - x$offset), p), ncol = p
+    ) * X1 # n by p, wic * residual * xi
+  
+  wi <- x$weights
+  X2 <- wi * x$residuals  # n by 1, wi[z] * (residual - rhoz) * z
+  
+  XX <- cbind(X1, X2)
+  XX <- aggregate(XX, by = list(cid), FUN = sum)[,-1]
+  
+  # multiple sqrt(product of #treated divided by block sizes) to XX by group
+  nbk <- design_table(design = design_obj, x = "treatment", y = "block")
+  const <- sqrt(nbk[, 1] * nbk[, 2] / rowSums(nbk))
+  XX <- sweep(XX, 1, const[bid], '*')
+  return(XX)
+}
+
+#' @title (Internal) Helper function for design-based meat matrix calculation
+#' @keywords internal
```

Comment:
Can you add a little more documentation here to explain what this function computes?

Reply:
Added documentation. See commit https://github.com/benbhansen-stats/propertee/pull/185/commits/3f27a1e6c731038c3d2dd81cbee8451fdfe5bbcc